File: csharp/patterns/tests/PrototypePatterns.Tests/PatternTests.cs
using PrototypePatterns.Behavioral.ChainOfResponsibility;
using PrototypePatterns.Behavioral.Command;
using PrototypePatterns.Behavioral.Iterator;
using PrototypePatterns.Behavioral.Mediator;
using PrototypePatterns.Behavioral.Memento;
using PrototypePatterns.Behavioral.Observer;
using PrototypePatterns.Behavioral.State;
using PrototypePatterns.Behavioral.Strategy;
using PrototypePatterns.Behavioral.TemplateMethod;
using PrototypePatterns.Creational.AbstractFactory;
using PrototypePatterns.Creational.Builder;
using PrototypePatterns.Creational.FactoryMethod;
using PrototypePatterns.Creational.LazyInitialization;
using PrototypePatterns.Creational.ObjectPool;
using PrototypePatterns.Creational.Prototype;
using PrototypePatterns.Creational.Singleton;
using PrototypePatterns.Structural.Adapter;
using PrototypePatterns.Structural.Bridge;
using PrototypePatterns.Structural.Composite;
using PrototypePatterns.Structural.CompositeEntity;
using PrototypePatterns.Structural.Decorator;
using PrototypePatterns.Structural.Facade;
using PrototypePatterns.Structural.Flyweight;
using PrototypePatterns.Structural.Proxy;

namespace PrototypePatterns.Tests;

public sealed class PatternTests
{
    [Fact]
    public void Singleton() => Assert.True(SingletonPattern.Demo());

    [Fact]
    public void FactoryMethod() => Assert.Equal(["<button>Submit</button>", "[DesktopButton: Submit]"], FactoryMethodPattern.Demo());

    [Fact]
    public void AbstractFactory() => Assert.Equal(["dark-button", "dark-checkbox"], AbstractFactoryPattern.Demo());

    [Fact]
    public void Builder()
    {
        var house = BuilderPattern.Demo();
        Assert.Equal(["Kitchen", "Office"], house.Rooms);
        Assert.True(house.HasGarage);
    }

    [Fact]
    public void Prototype() => Assert.True(PrototypePattern.Demo());

    [Fact]
    public void ObjectPool() => Assert.True(ObjectPoolPattern.Demo());

    [Fact]
    public void LazyInitialization() => Assert.True(LazyInitializationPattern.Demo());

    [Fact]
    public void Adapter() => Assert.Equal(25.0, AdapterPattern.Demo());

    [Fact]
    public void Decorator() => Assert.Equal("email:build complete|sms:build complete|slack:build complete", DecoratorPattern.Demo());

    [Fact]
    public void Composite() => Assert.Equal(30, CompositePattern.Demo());

    [Fact]
    public void Facade() => Assert.Contains("projector on", FacadePattern.Demo());

    [Fact]
    public void Proxy() => Assert.True(ProxyPattern.Demo());

    [Fact]
    public void Bridge() => Assert.Equal(["vector-circle:5", "raster-circle:5"], BridgePattern.Demo());

    [Fact]
    public void Flyweight() => Assert.True(FlyweightPattern.Demo());

    [Fact]
    public void CompositeEntity() => Assert.Equal(["Sam", "Austin"], CompositeEntityPattern.Demo());

    [Fact]
    public void Observer() => Assert.Equal([7], ObserverPattern.Demo());

    [Fact]
    public void Strategy() => Assert.Equal([[1, 2, 3], [3, 2, 1]], StrategyPattern.Demo());

    [Fact]
    public void Command() => Assert.Equal("light:on", CommandPattern.Demo());

    [Fact]
    public void State() => Assert.Equal(["pending->paid", "paid->shipped", "shipped->shipped"], StatePattern.Demo());

    [Fact]
    public void TemplateMethod() => Assert.Equal(["read:csv", "transform:normalize", "write:warehouse"], TemplateMethodPattern.Demo());

    [Fact]
    public void ChainOfResponsibility() => Assert.Equal(["team-lead", "manager", "unhandled:4"], ChainOfResponsibilityPattern.Demo());

    [Fact]
    public void Iterator() => Assert.Equal(["intro", "verse", "outro"], IteratorPattern.Demo());

    [Fact]
    public void Mediator() => Assert.Equal(["alex:ready"], MediatorPattern.Demo());

    [Fact]
    public void Memento() => Assert.Equal("draft", MementoPattern.Demo());
}

File: java/patterns/src/test/java/org/prototypepatterns/PatternTests.java
package org.prototypepatterns;

import java.util.List;
import org.prototypepatterns.behavioral.chainofresponsibility.ChainOfResponsibilityPattern;
import org.prototypepatterns.behavioral.command.CommandPattern;
import org.prototypepatterns.behavioral.iterator.IteratorPattern;
import org.prototypepatterns.behavioral.mediator.MediatorPattern;
import org.prototypepatterns.behavioral.memento.MementoPattern;
import org.prototypepatterns.behavioral.observer.ObserverPattern;
import org.prototypepatterns.behavioral.state.StatePattern;
import org.prototypepatterns.behavioral.strategy.StrategyPattern;
import org.prototypepatterns.behavioral.templatemethod.TemplateMethodPattern;
import org.prototypepatterns.creational.abstractfactory.AbstractFactoryPattern;
import org.prototypepatterns.creational.builder.BuilderPattern;
import org.prototypepatterns.creational.factorymethod.FactoryMethodPattern;
import org.prototypepatterns.creational.lazyinitialization.LazyInitializationPattern;
import org.prototypepatterns.creational.objectpool.ObjectPoolPattern;
import org.prototypepatterns.creational.prototypepattern.PrototypePattern;
import org.prototypepatterns.creational.singleton.SingletonPattern;
import org.prototypepatterns.structural.adapter.AdapterPattern;
import org.prototypepatterns.structural.bridge.BridgePattern;
import org.prototypepatterns.structural.composite.CompositePattern;
import org.prototypepatterns.structural.compositeentity.CompositeEntityPattern;
import org.prototypepatterns.structural.decorator.DecoratorPattern;
import org.prototypepatterns.structural.facade.FacadePattern;
import org.prototypepatterns.structural.flyweight.FlyweightPattern;
import org.prototypepatterns.structural.proxy.ProxyPattern;

public final class PatternTests {
    private static int passed = 0;

    private PatternTests() {
    }

    private static void assertTrue(boolean condition, String message) {
        if (!condition) {
            throw new IllegalStateException(message);
        }
        passed++;
    }

    private static void assertEquals(Object expected, Object actual, String message) {
        assertTrue(expected.equals(actual), message + " expected=" + expected + " actual=" + actual);
    }

    private static void assertDouble(double expected, double actual, String message) {
        assertTrue(Math.abs(expected - actual) < 0.0001, message + " expected=" + expected + " actual=" + actual);
    }

    public static void main(String[] args) {
        assertTrue(SingletonPattern.demo(), "singleton");
        assertEquals(List.of("<button>Submit</button>", "[DesktopButton: Submit]"), FactoryMethodPattern.demo(), "factory method");
        assertEquals(List.of("dark-button", "dark-checkbox"), AbstractFactoryPattern.demo(), "abstract factory");
        BuilderPattern.House house = BuilderPattern.demo();
        assertEquals(List.of("Kitchen", "Office"), house.rooms, "builder rooms");
        assertTrue(house.hasGarage, "builder garage");
        assertTrue(PrototypePattern.demo(), "prototype");
        assertTrue(ObjectPoolPattern.demo(), "object pool");
        assertTrue(LazyInitializationPattern.demo(), "lazy initialization");

        assertDouble(25.0, AdapterPattern.demo(), "adapter");
        assertEquals("email:build complete|sms:build complete|slack:build complete", DecoratorPattern.demo(), "decorator");
        assertTrue(CompositePattern.demo() == 30, "composite");
        assertTrue(FacadePattern.demo().contains("projector on"), "facade");
        assertTrue(ProxyPattern.demo(), "proxy");
        assertEquals(List.of("vector-circle:5", "raster-circle:5"), BridgePattern.demo(), "bridge");
        assertTrue(FlyweightPattern.demo(), "flyweight");
        assertEquals(List.of("Sam", "Austin"), CompositeEntityPattern.demo(), "composite entity");

        assertEquals(List.of(7), ObserverPattern.demo(), "observer");
        assertEquals(List.of(List.of(1, 2, 3), List.of(3, 2, 1)), StrategyPattern.demo(), "strategy");
        assertEquals("light:on", CommandPattern.demo(), "command");
        assertEquals(List.of("pending->paid", "paid->shipped", "shipped->shipped"), StatePattern.demo(), "state");
        assertEquals(List.of("read:csv", "transform:normalize", "write:warehouse"), TemplateMethodPattern.demo(), "template method");
        assertEquals(List.of("team-lead", "manager", "unhandled:4"), ChainOfResponsibilityPattern.demo(), "chain of responsibility");
        assertEquals(List.of("intro", "verse", "outro"), IteratorPattern.demo(), "iterator");
        assertEquals(List.of("alex:ready"), MediatorPattern.demo(), "mediator");
        assertEquals("draft", MementoPattern.demo(), "memento");

        assertTrue(PatternRegistry.list().size() == 24, "registry size");
        assertEquals(true, PatternRegistry.runPattern("singleton"), "registry singleton");
        assertEquals(List.of("team-lead", "manager", "unhandled:4"), PatternRegistry.runPattern("chain of responsibility"),
                "registry normalized id");
        assertTrue(PatternRegistry.runAllPatterns().size() == 24, "registry all");

        System.out.println("Java pattern tests passed: " + passed);
    }
}
